### PR TITLE
Base: Parse Rumors

### LINKED ISSUE #2: Parse: rumours

### PR DESCRIPTION
**Pull request by san-e**

### CODE REVIEW

File: flint/entities/universe.py [hunk 1]
```diff
@@ -101,6 +101,26 @@ def mbase(self) -> Optional[missions.MBase]:
...
+                if type(npc.rumor) != list: npc.rumor = [npc.rumor]
+                for rumor in npc.rumor:
+                    temp.append(lookup(rumor[3]))
+                if temp != []:
+                    if not rumors.get(npc.affiliation):
+                        rumors[npc.affiliation] = temp
+                    else:
+                        rumors[npc.affiliation] = rumors.get(npc.affiliation) + temp
+            
+            return {affiliation: list(dict.fromkeys(content)) for affiliation, content in rumors.items()}
```
Comment (biqqles) [inline flint/entities/universe.py]: Is `list(dict.fromkeys(content))` to remove dups? You could just return a set instead.

File: flint/entities/universe.py [hunk 2]
```diff
@@ -101,6 +101,26 @@ def mbase(self) -> Optional[missions.MBase]:
...
+    def rumors(self, markup = "html") -> dict:
+        """All rumors offered on this base, of the form {faction -> rumors}"""
+        lookup = self._markup_formats[markup]
+        if self.mbase():
+            rumors = {}
+            npcs = self.mbase().npcs
+
+            for npc in npcs:
+                temp = []
+                if type(npc.rumor) != list: npc.rumor = [npc.rumor]
```
Comment (biqqles) [inline flint/entities/universe.py]: Split to multiple lines

File: flint/entities/universe.py [hunk 3]
```diff
@@ -101,6 +101,26 @@ def mbase(self) -> Optional[missions.MBase]:
         """The mission base entry for this base."""
         return missions.get_mbases().get(self.nickname)
 
+    def rumors(self, markup = "html") -> dict:
+        """All rumors offered on this base, of the form {faction -> rumors}"""
```
Comment (biqqles) [inline flint/entities/universe.py:107]:
```suggestion
        """All rumors offered on this base, of the form {faction -> rumors}."""
```

File: flint/entities/universe.py [hunk 4]
```diff
@@ -101,6 +101,26 @@ def mbase(self) -> Optional[missions.MBase]:
         """The mission base entry for this base."""
         return missions.get_mbases().get(self.nickname)
 
+    def rumors(self, markup = "html") -> dict:
```
Comment (biqqles) [inline flint/entities/universe.py]: Only a valid type annotation in newer Python versions, we support back to 3.6

File: flint/entities/universe.py (hunk 1 above)
Comment (biqqles) [inline flint/entities/universe.py]: Should probably have `Faction` objects as keys

**Review by biqqles — CHANGES_REQUESTED** [review]
Thanks, this is a great feature! Some comments, might make some more later.

**Comment by biqqles** [comment]
I think my refactor should give equivalent output to yours, please let me know if I missed anything.

**Comment by san-e** [comment]
Hi, thanks for the refactor! It produces identical output most of the time, but sometimes returns an empty `set()` object. Should be fixed by just adding a `if npc.rumor:` statement to the top of the npc loop

**Comment by biqqles** [comment]
Thanks!

**Pull request merged**